cdap-metadata-ext-spanner/src/main/java/io/cdap/cdap/metadata/spanner/SpannerMetadataStorage.java

```
@@ -779,7 +790,287 @@ protected String toMetadataId(MetadataEntity entity) {
 
   @Override
   public SearchResponse search(SearchRequest request) throws IOException {
-    throw new IOException("NOT IMPLEMENTED");
+    return request.getCursor() != null && !request.getCursor().isEmpty()
+      ? doScroll(request) : doSearch(request);
+  }
+  private io.cdap.cdap.spi.metadata.SearchResponse doScroll(SearchRequest request) throws IOException {
+    Cursor cursor = Cursor.fromString(request.getCursor());
+    return performSpannerSearch(request, cursor);
+  }
+
+  private io.cdap.cdap.spi.metadata.SearchResponse doSearch(SearchRequest request) throws IOException {
+    return performSpannerSearch(request, null);
+  }
+
+  private io.cdap.cdap.spi.metadata.SearchResponse performSpannerSearch(SearchRequest request,
+                                                                        @Nullable Cursor cursor) throws IOException {
+    try (ReadOnlyTransaction transaction = dbClient.readOnlyTransaction()) {
+
+      String sql = buildSpannerQuery(request, cursor);
+      Statement statement = Statement.newBuilder(sql).build();
+      ResultSet resultSet = transaction.executeQuery(statement);
+
+      LOG.info(sql);
+      LOG.info(statement.toString());
+      LOG.info(resultSet.toString());
+
+      List<MetadataRecord> results = new ArrayList<>();
+      String nextCursor = null;
+
+      while (resultSet.next()) {
+        results.add(mapSpannerResult(resultSet));
+        nextCursor = createNextCursorKey(resultSet);
+      }
+
+      if (results.isEmpty()) {
+        nextCursor = null;
+      }
+      LOG.info(results.toString());
+      return createSpannerSearchResponse(request, results, nextCursor);
+    } catch (SpannerException e) {
+      throw new IOException("Spanner search failed", e);
+    }
```

this is the internal function is it okay to let the function throw original exception and the caller should wrap it as needed?